prusti-encoder/src/encoders/mir_pure.rs

```diff
@@ -814,6 +814,10 @@ impl<'vir: 'enc, 'enc> Enc<'vir, 'enc> {
                             self.deps.require_ref::<CastToEnc<CastTypePure>>(cast_args)
                                 .unwrap().apply_cast_if_necessary(self.vcx, proj_app)
 
+                        } else if let TyKind::Tuple(_) = tykind {
+                            self
+                                .deps.require_local::<RustTyCastersEnc<CastTypePure>>(ty)
+                                .unwrap().cast_to_concrete_if_possible(self.vcx, proj_app)
```

Comment:
@zgrannan Does this look right to you? Without this change I get a type error if a Rust tuple is used in pure code. More specifically, I tried a postcondition with a checked binop `#[ensures(result == a + b)]` (I also made an `Assert` behave like a `Goto` in the pure encoder, otherwise there is another panic). It seems a bit messy to me that this should be a special case here at a callsite rather than something hidden in the casters.

Reply:
Yeah this looks right, and I agree this logic should not be here and instead pushed into its own encoder.